packages/itwin/grouping-mapping-widget/src/widget/components/CustomCalculationAction.tsx

```diff
@@ -9,51 +9,128 @@ import {
+  onSaveSuccess: () => void;
+  onClickCancel?: () => void;
 }
 
-const CustomCalculationAction = ({
-  iModelId,
+const stringToPossibleDataType = (str?: string): PossibleDataType => {
+  if (!str)
+    return "Undefined";
+
+  switch (str.toLowerCase()) {
+    case "double":
+    case "number": return "Number";
+    case "string": return "String";
+    case "boolean": return "Boolean";
+    default: return "Undefined";
+  }
+};
+
+const convertToPropertyMap = (
+  groupProperties: GroupProperty[],
+  calculatedProperties: CalculatedProperty[],
+  customCalculations: CustomCalculation[],
+  selectedPropertyName?: string
+): PropertyMap => {
+  const map: PropertyMap = {};
+  const selectedLowerName = selectedPropertyName?.toLowerCase();
+
+  groupProperties.forEach((p) => {
+    const lowerName = p.propertyName?.toLowerCase();
+    if (lowerName && lowerName !== selectedLowerName)
+      map[lowerName] = stringToPossibleDataType(p.dataType);
+  });
+
+  calculatedProperties.forEach((p) => {
+    const lowerName = p.propertyName?.toLowerCase();
+    if (lowerName)
+      map[lowerName] = "Number";
+  });
+
+  customCalculations.forEach((p) => {
+    const lowerName = p.propertyName?.toLowerCase();
+    if (lowerName && lowerName !== selectedLowerName)
+      map[lowerName] = stringToPossibleDataType(p.dataType);
+  });
+
+  return map;
+};
+
+export const CustomCalculationAction = ({
   mappingId,
   groupId,
-  properties,
   customCalculation,
-  returnFn,
-}: CalculatedPropertyActionProps) => {
-  const { getAccessToken } = useGroupingMappingApiConfig();
+  onSaveSuccess,
+  onClickCancel,
+}: CustomCalculationActionProps) => {
+  const { getAccessToken, iModelId } = useGroupingMappingApiConfig();
   const mappingClient = useMappingClient();
   const [propertyName, setPropertyName] = useState<string>(
     customCalculation?.propertyName ?? "",
   );
   const [formula, setFormula] = useState<string>(
     customCalculation?.formula ?? "",
   );
+  const {
+    groupProperties,
+    setGroupProperties,
+    calculatedProperties,
+    setCalculatedProperties,
+    customCalculationProperties,
+    setCustomCalculationProperties,
+  } = usePropertiesContext();
   const [quantityType, setQuantityType] = useState<QuantityType>(customCalculation?.quantityType ?? QuantityType.Undefined);
-  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const [isLoading, setIsLoading] = useState<boolean>(true);
   const [formulaErrorMessage, setFormulaErrorMessage] = useState<string>("");
   const [validator, showValidationMessage] = useValidator();
+  const [properties, setProperties] = useState<PropertyMap>({});
   const { isValid, forceValidation } = useFormulaValidation(propertyName.toLowerCase(), formula, properties, setFormulaErrorMessage);
 
+  const initialize = useCallback(async () => {
+    const groupProperties = await mappingClient.getGroupProperties((await getAccessToken()), iModelId, mappingId, groupId);
```

Comment:
Querying of properties should be done in parallel. There is no point doing them in sequence.